### PR TITLE
added support for single game state file

### PR DESCRIPTION
**Pull request by weseldridge**
This deals with #72. 

~~I removed the JSON save from the render so you can save game state without rendering the game.~~ also added a util to merge all of the final game states into one file. 

~~There are also some changes in the `run_battle` to support the removal of saving JSON files from the render method.~~

### CODE REVIEW

File: pommerman/cli/run_battle.py [hunk 1]
```diff
@@ -55,8 +57,10 @@ def _run(seed, record_pngs_dir=None, record_json_dir=None):
             steps += 1
             if args.render:
                 env.render(record_pngs_dir=args.record_pngs_dir,
-                           record_json_dir=args.record_json_dir,
                            mode=args.render_mode)
+            if record_json_dir is not None:
```
Comment (cinjon) [inline pommerman/cli/run_battle.py]: why take this out of the env.render and then put it here?

File: pommerman/utility.py [hunk 2]
```diff
@@ -326,3 +328,34 @@ def get_next_position(position, direction):
...
+        "agents": agents,
+        "finishedAt": finishedAt,
+        "config": config,
+        "state": []
+    }
+
+    merger = Merger(josnSchema)
+    base = merger.merge({}, jsonTemplate)
+
+    for root, dirs, files in os.walk(record_json_dir):
```
Comment (cinjon) [inline pommerman/utility.py]: does this preserve the order?

File: pommerman/envs/v0.py [hunk 3]
```diff
@@ -282,6 +277,14 @@ def featurize(obs):
             (board, bomb_blast_strength, bomb_life, position, ammo,
              blast_strength, can_kick, teammate, enemies))
 
+    def save_json(self, record_json_dir):
+        info = self.get_json_info()
+        count = "{0:0=3d}".format(self._step_count)
+        suffix = count + '.json'
+        path = os.path.join(record_json_dir, suffix)
+        with open(path, 'w') as f:
+                f.write(json.dumps(info, sort_keys=True, indent=4))
```
Comment (cinjon) [inline pommerman/envs/v0.py]: indented too far

File: pommerman/utility.py [hunk 4]
```diff
@@ -326,3 +328,34 @@ def get_next_position(position, direction):
 
 def make_np_float(feature):
     return np.array(feature).astype(np.float32)
+
+def join_json_state(record_json_dir, agents, finishedAt, config):
+    josnSchema = {
```
Comment (cinjon) [inline pommerman/utility.py]: name

File: pommerman/utility.py (hunk 2 above)
Comment (weseldridge) [reply pommerman/utility.py]: indeed.

File: pommerman/envs/v0.py (hunk 3 above)
Comment (weseldridge) [reply pommerman/envs/v0.py]: 👌

File: pommerman/cli/run_battle.py (hunk 1 above)
Comment (weseldridge) [reply pommerman/cli/run_battle.py]: So when it runs on the server the game doesn't have to render.

File: pommerman/utility.py (hunk 4 above)
Comment (weseldridge) [reply pommerman/utility.py]: typical. 😹It wouldn't be a proper PR if I didn't misspell something.

File: pommerman/cli/run_battle.py (hunk 1 above)
Comment (cinjon) [reply pommerman/cli/run_battle.py]: I'm a bit confused by that. The same machine that's rendering here is the one that's simulating. This isn't parallelized or being sent over a wire.

Comment (weseldridge) [reply pommerman/cli/run_battle.py]: I just thought we wouldn't be calling render when running sims on the server. I do not mind moving it back.

File: pommerman/cli/run_battle.py [hunk 5]
```diff
@@ -46,7 +48,7 @@ def run(args, num_times=1, seed=None):
     if record_json_dir and not os.path.isdir(record_json_dir):
         os.makedirs(record_json_dir)
 
-    def _run(seed, record_pngs_dir=None, record_json_dir=None):
```
Comment (cinjon) [inline pommerman/cli/run_battle.py]: why did you remove the record_json_dir argument?

Comment (weseldridge) [reply pommerman/cli/run_battle.py]: It is defined in the file's global scope and b/c of the way it works now there is no reason to create subdirectories for each iteration. The merge JSON function will create a time-stamped file like `game_state_20180528091945.json`. It will then remove all of the old single step files.

File: pommerman/cli/run_battle.py [hunk 6]
```diff
@@ -71,6 +73,12 @@ def _run(seed, record_pngs_dir=None, record_json_dir=None):
             if do_sleep:
                 time.sleep(5)
             env.render(close=True)
+
+        if record_json_dir:
+            finishedAt = datetime.now().isoformat()
```
Comment (cinjon) [inline pommerman/cli/run_battle.py]: finishedAt --> finished_at

File: pommerman/utility.py [hunk 7]
```diff
@@ -328,3 +331,43 @@ def get_next_position(position, direction):
...
+        "properties": {
+            "state": {
+                "mergeStrategy": "append"
+            }
+        }
+    }
+
+    jsonTemplate = {
+        "agents": agents,
+        "finishedAt": finishedAt,
```
Comment (cinjon) [inline pommerman/utility.py]: s/finishedAt/finished_at

File: pommerman/utility.py [hunk 8]
```diff
@@ -328,3 +331,43 @@ def get_next_position(position, direction):
...
+    for root, dirs, files in os.walk(record_json_dir):
+        for name in files:
+            path = os.path.join(record_json_dir, name)
+            if name.endswith('.json') and "game_state" not in name:
+                with open(path) as data_file:    
+                    data = json.load(data_file)
+                    head = {"state":[data]}
+                    base = merger.merge(base, head)
+    
+    filename = 'game_state_' + datetime.now().strftime('%Y%m%d%H%M%S') + '.json'
```
Comment (cinjon) [inline pommerman/utility.py]: what's the different between this datetime and finished_at?

File: pommerman/utility.py (hunk 7 above)
Comment (weseldridge) [reply pommerman/utility.py]: cool. an old habit of working with mongo.

File: pommerman/utility.py (hunk 8 above)
Comment (weseldridge) [reply pommerman/utility.py]: Nothing. TBH if each episode is in it's own sub dir then we can remove it.
